dns/cache/dns_cache.go

```diff
@@ -0,0 +1,60 @@
+package cache
+
+import (
+	"sync"
+	"time"
+
+	"github.com/xvzc/SpoofDPI/util"
+)
+
+type DNSCacheEntry struct {
+	ip           string
+	expiry_timer *time.Timer
+}
+
+type DNSCache struct {
+	cacheMap  map[string]DNSCacheEntry
+	cacheLock sync.RWMutex
+}
+
+var dnsCache *DNSCache
+var once sync.Once
+
+func GetCache() *DNSCache {
+	once.Do(func() {
+		dnsCache = &DNSCache{
+			cacheMap:  make(map[string]DNSCacheEntry),
+			cacheLock: sync.RWMutex{},
+		}
+	})
+	return dnsCache
+}
+
+// interface function for the inner map basically.
+func (d *DNSCache) Set(key string, value string) {
+	d.cacheLock.Lock()
+	defer d.cacheLock.Unlock()
+	if _, ok := d.cacheMap[key]; ok {
+		return
+	}
+
+	d.cacheMap[key] = DNSCacheEntry{
+		ip: value,
+		expiry_timer: time.AfterFunc(time.Duration(*&util.GetConfig().DnsCacheTTL)*time.Second, func() {
```

Comment:
Is it possible to do this without a daemon routine? For example with lazy evict, when getting the cache entry, compare the current time and the time the record was created. I don't know how much overhead this adds to each record.